Mathlib/RingTheory/NoetherNormalization.lean

```diff
@@ -0,0 +1,401 @@
+/-!
+# Norther Normalizaton
+This file contains a proof of Noether normalization lemma by Nagata.
+
+## Main Results
+Let $A$ be a finitely generated algebra over a field $k$.
+Then there exists natural number $r$ and an injective homomorphism
+from $k[X_1, X_2, ..., X_r]$ to $A$ such that $A$ is integral over $k[X_1, X_2, ..., X_r]$.
+
+## Reference
+* <https://stacks.math.columbia.edu/tag/00OW>
+
+-/
+open Polynomial MvPolynomial Ideal BigOperators Nat RingHom
+
+variable {k : Type*} [Field k] {n : ℕ} (f : MvPolynomial (Fin (n + 1)) k)
+variable (v w : Fin (n + 1) →₀ ℕ)
+
+/-Suppose $f$ is a nonzero polynomial of $n+1$ variables : $X_0,...,X_n$.
+We construct a ring homomorphism $T$ which maps $X_i$ into $X_i + X_0^{r_i}$ when $i \neq 0$,
+and maps $f$ into some monic polynomial (regarded as a polynomial of $X_0$).-/
+private noncomputable abbrev up := (1 + ∑ i : Fin (n + 1), f.degreeOf i)
+
+private lemma up_spec (hv : v ∈ f.support) : ∀ i, v i < up f := by
+  intro i
+  have h : degreeOf i f ≤ ∑ i : Fin (n + 1), degreeOf i f := by
+    rw [← finsum_eq_sum_of_fintype fun i ↦ degreeOf i f]
+    exact single_le_finsum i (Set.toFinite _) (fun j ↦ Nat.zero_le (degreeOf j f))
+  exact lt_one_add_iff.mpr (le_trans (monomial_le_degreeOf i hv) h)
+
+private noncomputable abbrev r : Fin (n + 1) → ℕ :=
+    fun i ↦ if i = 0 then 1 else ((n + 1) * (up f)) ^ (n + 1 - i)
+
+/-$T$ is an algebraic isomorphism.-/
+private noncomputable abbrev T1 : MvPolynomial (Fin (n + 1)) k →ₐ[k] MvPolynomial (Fin (n + 1)) k :=
+  aeval (fun i ↦  if i = 0 then X 0 else X i + (X 0) ^ (r f i))
+
+private noncomputable abbrev T_inv :
+    MvPolynomial (Fin (n + 1)) k →ₐ[k] MvPolynomial (Fin (n + 1)) k :=
+  aeval (fun i ↦  if i = 0 then X 0 else X i - (X 0) ^ (r f i))
+
+private lemma T_left_inv : (T_inv f).comp (T1 f) = AlgHom.id _ _ := by
+  rw [comp_aeval, ← MvPolynomial.aeval_X_left]
+  congr
+  ext i v
+  cases' i using Fin.cases with d
+  · simp only [if_pos, MvPolynomial.aeval_X]
+  · simp only [Fin.succ_ne_zero, ite_false, map_add, map_pow,
+      MvPolynomial.aeval_X, if_pos, sub_add_cancel]
+
+private lemma T_right_inv : (T1 f).comp (T_inv f) = AlgHom.id _ _ := by
+  rw [comp_aeval, ← MvPolynomial.aeval_X_left]
+  congr
+  ext i v
+  cases' i using Fin.cases with d
+  · simp only [if_pos, MvPolynomial.aeval_X]
+  · simp only [Fin.succ_ne_zero, ite_false, map_sub, map_pow,
+      MvPolynomial.aeval_X, if_pos, add_sub_cancel_right]
+
+private noncomputable abbrev T := AlgEquiv.ofAlgHom (T1 f) (T_inv f) (T_right_inv f) (T_left_inv f)
+
+/-$r_i$ satisfies that any two different monomial in $f$ have different degree of $X_0$
+after mapped by $T$.-/
+private lemma a_j (vlt : ∀ i, v i < up f) (wlt : ∀ i, w i < up f) (i : Fin n)
+  (pj : ∀ (j : Fin n), (v j.succ ≠ w j.succ) → i ≤ j) :  ∀ j : Fin n, (if j ≠ i then
+  |(((n + 1 : ℤ) * up f) ^ (n + 1 - j.succ) * ((v j.succ) - (w j.succ)))| else 0)
+  ≤ (n + 1 : ℤ) ^ (n - i.succ) * up f ^ (n + 1 - i.succ) := by
```

Comment:
Have you looked into [Nat.ofDigits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Digits.html#Nat.ofDigits)? [Nat.digits_ofDigits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Digits.html#Nat.digits_ofDigits) should give the injectivity easily. There is [List.ofFn](https://leanprover-community.github.io/mathlib4_docs/Batteries/Data/List/Basic.html#List.ofFn) to convert a Fin tuple to a list.

It's not clear to me what this lemma is doing ...